File: backend_thien_tam_app/src/models/BookCategory.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IBookCategory extends Document {
  _id: string;
  name: string;
  nameEn?: string;
  description?: string;
  icon?: string;
  color?: string;
  displayOrder: number;
  isActive: boolean;
  bookCount: number;
  createdAt: Date;
  updatedAt: Date;
}

const bookCategorySchema = new Schema<IBookCategory>(
  {
    name: {
      type: String,
      required: [true, 'Tên danh mục là bắt buộc'],
      unique: true,
      trim: true,
    },
    nameEn: {
      type: String,
      trim: true,
    },
    description: {
      type: String,
      trim: true,
    },
    icon: {
      type: String,
      default: '📚',
    },
    color: {
      type: String,
      default: '#8B7355',
    },
    displayOrder: {
      type: Number,
      default: 0,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    bookCount: {
      type: Number,
      default: 0,
    },
  },
  {
    timestamps: true,
  }
);

// Index for sorting
bookCategorySchema.index({ displayOrder: 1, name: 1 });
bookCategorySchema.index({ isActive: 1 });

export const BookCategory = mongoose.model<IBookCategory>(
  'BookCategory',
  bookCategorySchema
);

File: thien_tam_app/lib/features/books/data/models/book_category.dart
class BookCategory {
  final String id;
  final String name;
  final String? nameEn;
  final String? description;
  final String icon;
  final String color;
  final int displayOrder;
  final bool isActive;
  final int bookCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  const BookCategory({
    required this.id,
    required this.name,
    this.nameEn,
    this.description,
    required this.icon,
    required this.color,
    required this.displayOrder,
    required this.isActive,
    required this.bookCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory BookCategory.fromJson(Map<String, dynamic> json) {
    return BookCategory(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      nameEn: json['nameEn'],
      description: json['description'],
      icon: json['icon'] ?? '📚',
      color: json['color'] ?? '#8B7355',
      displayOrder: (json['displayOrder'] as num?)?.toInt() ?? 0,
      isActive: json['isActive'] ?? true,
      bookCount: (json['bookCount'] as num?)?.toInt() ?? 0,
      createdAt: DateTime.parse(
        json['createdAt'] ?? DateTime.now().toIso8601String(),
      ),
      updatedAt: DateTime.parse(
        json['updatedAt'] ?? DateTime.now().toIso8601String(),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'nameEn': nameEn,
      'description': description,
      'icon': icon,
      'color': color,
      'displayOrder': displayOrder,
      'isActive': isActive,
      'bookCount': bookCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  // For creating/updating category
  Map<String, dynamic> toCreateJson() {
    return {
      'name': name,
      if (nameEn != null) 'nameEn': nameEn,
      if (description != null) 'description': description,
      'icon': icon,
      'color': color,
      'displayOrder': displayOrder,
      'isActive': isActive,
    };
  }

  BookCategory copyWith({
    String? id,
    String? name,
    String? nameEn,
    String? description,
    String? icon,
    String? color,
    int? displayOrder,
    bool? isActive,
    int? bookCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return BookCategory(
      id: id ?? this.id,
      name: name ?? this.name,
      nameEn: nameEn ?? this.nameEn,
      description: description ?? this.description,
      icon: icon ?? this.icon,
      color: color ?? this.color,
      displayOrder: displayOrder ?? this.displayOrder,
      isActive: isActive ?? this.isActive,
      bookCount: bookCount ?? this.bookCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() {
    return 'BookCategory(id: $id, name: $name, nameEn: $nameEn, icon: $icon, bookCount: $bookCount)';
  }
}
